src/main/java/com/aws/greengrass/builtin/services/pubsub/SubscriptionTrie.java

```diff
@@ -0,0 +1,179 @@
+     */
+    public SubscriptionTrie() {
+        this("");
+    }
+
+    private SubscriptionTrie(String value) {
+        this.value = value;
+        this.callbacks = ConcurrentHashMap.newKeySet();
+    }
+
+    private SubscriptionTrie lookup(String topic) {
+        SubscriptionTrie current = this;
+        for (String topicLevel : topic.split(String.valueOf(TOPIC_LEVEL_SEPARATOR))) {
+            current = current.children.get(topicLevel);
+            if (current == null) {
+                return null;
+            }
+        }
+        return current;
+    }
+
+    /**
+     * Check if trie contains a topic.
+     *
+     * @param topic topic
+     * @return if trie contains topic
+     */
+    public boolean containsKey(String topic) {
+        return lookup(topic) != null;
+    }
+
+    /**
+     * Remove entry for one callback.
+     *
+     * @param topic topic
+     * @param cb    callback
+     * @return if changed after removal
+     */
+    public boolean remove(String topic, Object cb) {
+        return remove(topic, Collections.singleton(cb));
+    }
+
+    /**
+     * Remove entry for a set of callbacks.
+     *
+     * @param topic topic
+     * @param cbs   callbacks
+     * @return if changed after removal
+     */
+    public boolean remove(String topic, Set<Object> cbs) {
+        SubscriptionTrie sub = lookup(topic);
+        if (sub == null) {
+            return false;
+        }
+        return sub.callbacks.removeAll(cbs);
+    }
+
+    /**
+     * Size of trie.
+     *
+     * @return size
+     */
+    public int size() {
+        int[] size = {this.callbacks.size()};
+        children.forEach((s, t) -> {
+            size[0] += t.size();
+        });
+        return size[0];
+    }
+
+    /**
+     * Add a topic callback.
+     *
+     * @param topic topic
+     * @param cb    callback
+     * @return true
+     */
+    public boolean add(String topic, Object cb) {
+        ConcurrentHashMap.KeySetView<Object, Boolean> cbs = ConcurrentHashMap.newKeySet();
+        cbs.add(cb);
+        put(topic, cbs);
+        return true;
+    }
+
+    /**
+     * Add a topic and a set of callbacks.
+     *
+     * @param topic topic
+     * @param cbs   callbacks
+     */
+    public void put(String topic, Set<Object> cbs) {
+        SubscriptionTrie[] current = {this};
+        for (String topicLevel : topic.split(String.valueOf(TOPIC_LEVEL_SEPARATOR))) {
+            current[0] = current[0].children.computeIfAbsent(topicLevel, SubscriptionTrie::new);
+        }
+        current[0].callbacks.addAll(cbs);
+    }
+
+    private Set<SubscriptionTrie> getMatchingPaths(String topicLevel, Set<Object> result) {
+        Set<SubscriptionTrie> paths = new LinkedHashSet<>();
+
+        SubscriptionTrie childPath = this.children.get(topicLevel);
+        if (childPath != null) {
+            paths.add(childPath);
+        }
+
+        SubscriptionTrie childPlusPath = this.children.get(String.valueOf(SINGLE_LEVEL_WILDCARD));
```

Comment:
Store these as both strings and characters so that you have the proper type already created; calling String.valueOf on every single lookup will add up and hurt performance.

`final SINGLE_LEVEL_WILDCARD_STR = "+";`
`final SINGLE_LEVEL_WILDCARD = SINGLE_LEVEL_WILDCARD_STR.charAt(0);`